semgrep-core/toy-matcher/Matcher.ml

```diff
@@ -0,0 +1,562 @@
+    | Some env -> Some (Env.bindings env)
+  in
+  opt_captures, stat
+
+(********** Tests **********)
+
+let print_time name f =
+  let t1 = Unix.gettimeofday () in
+  let res = f () in
+  let t2 = Unix.gettimeofday () in
+  printf "%s: %.6f s\n%!" name (t2 -. t1);
+  res
+
+let check_match ?cache pat input_str expected_opt_bindings =
+  let sort = Option.map (List.sort compare) in
+  let normalize opt_bindings =
+    Option.map (fun bindings ->
+      List.map (fun (var, symbols) -> (var, unparse symbols)) bindings
+    ) opt_bindings
+    |> sort
+  in
+  check_pattern pat;
+  let input = parse input_str in
+  let expected = sort expected_opt_bindings in
+  let actual =
+    let res, stat =
+      print_time "match function" (fun () ->
+        match_input ?cache pat input
+      )
+    in
+    printf "input length: %i\n" (String.length input_str);
+    printf "number of calls to the match function: %i\n" stat.match_calls;
+    normalize res
+  in
+  print_result stdout actual;
+  Alcotest.(check bool) "equal" true (expected = actual)
+
+let test_simple_symbol () =
+  let pat = [Symbol 'A', Some "a"] in
+  let input = "A" in
+  check_match pat input (Some ["a", "A"])
+
+let test_simple_ellipsis () =
+  let pat = [Ellipsis, Some "x"] in
+  let input = "ABC" in
+  check_match pat input (Some ["x", "ABC"])
+
+let test_any_symbol () =
+  let pat = [Any_symbol, Some "a"] in
+  let input = "A" in
+  check_match pat input (Some ["a", "A"])
+
+let test_floating_symbol () =
+  let pat = [
+    Ellipsis, Some "head";
+    Symbol 'A', Some "a";
+    Ellipsis, Some "tail";
+  ] in
+  let input = "012345A6789" in
+  check_match pat input (Some [
+    "head", "012345";
+    "a", "A";
+    "tail", "6789";
+  ])
+
+let test_backref ~cache () =
+  let pat = [
+    Ellipsis, None;
+    Any_symbol, Some "orig";
+    Backref "orig", Some "copy";
+    Ellipsis, None;
+  ] in
+  let input = "ABBC" in
+  check_match ~cache pat input (Some [
+    "orig", "B";
+    "copy", "B";
+  ])
+
+let test_gap () =
+  let pat = [
+    Symbol 'A', Some "a";
+    Ellipsis, Some "gap";
+    Symbol 'B', Some "b";
+  ] in
+  let input = "A12B" in
+  check_match pat input (Some [
+    "a", "A";
+    "gap", "12";
+    "b", "B";
+  ])
+
+let test_backref_backtrack ~cache () =
+  let pat = [
+    Ellipsis, None;
+    Any_symbol, Some "x";
+    Ellipsis, None;
+    Backref "x", None;
+    Symbol 'C', None;
+    Ellipsis, None;
+  ] in
+  let input = "ABBCA" in
+  check_match ~cache pat input (Some [
+    "x", "B";
+  ])
+
+(*
+   Deterministically generate a random-looking string.
+*)
+let pseudo_random_string len pick_from =
+  let n = String.length pick_from in
+  assert (n > 0);
+  Random.init 0;
+  String.init len (fun _i ->
+    pick_from.[Random.int n]
+  )
+
+(*
+   This is equivalent in semgrep to searching for a pattern like
+   '$A; ... foo;', in a file where the statement 'foo;' doesn't exist.
+
+   With the naive match algorithm, the complexity is O(n^2) where n is the
```

Comment:
can you explain why it's n^2, by unfolding the work done each time.

Reply:
I added a bunch of comments to the file. For this specifically, I tried explaining things in comments but concluded it's better to run `make bench` and observe the traces.